File: balloon.py
import hashlib

hashf = lambda *args: hashlib.sha256(''.join([str(arg) for arg in args]).encode()).hexdigest()
input_data = "some_random_input_data"
salt = 'some_random_salt'

def balloon(input_data, salt):
	space_cost = 16
	time_cost = 20
	delta = 3
	cnt = 0

	# Step 1. Expand input into buffer
	buf = [hashf(cnt, input_data, salt)]
	cnt += 1

	for m in range(1, space_cost):
		buf.append(hashf(cnt, buf[m - 1]))
		cnt += 1

	# Step 2. Mix buffer contents
	for t in range(time_cost):
		for m in range(space_cost):
			# Step 2a. Hash last and current blocks
			buf[m] = hashf(cnt, buf[(m - 1) % space_cost], buf[m])
			cnt += 1

			for i in range(delta):
				# Step 2b. Hash in pseudorandomly chosen blocks
				other = int(hashf(cnt, salt, t, m, i), 16) % space_cost
				cnt += 1
				buf[m] = hashf(cnt, buf[m], buf[other])
				cnt += 1

	# Step 3. Extract output from buffer
	return buf[-1]

print(balloon(input_data, salt))

File: balloon.c
#include "balloon.h"
#include "sphlib/sph_groestl.h"

#include <stdlib.h>
#include <stdint.h>
#include <string.h>

uint64_t u8tou64(uint8_t const* u8){
	uint64_t u64;
	memcpy(&u64, u8, sizeof(u64));
	return u64;
}

void balloon(const unsigned char* input, char* output, int length, const unsigned char* salt, int salt_length)
{
	const uint64_t s_cost = 64;
	const uint64_t t_cost = 2;
	const int delta = 3;
	
	sph_groestl512_context ctx_groestl;
	uint8_t blocks[s_cost][64];
	
	// Step 1: Expand input into buffer
	uint64_t cnt = 0;
	sph_groestl512_init(&ctx_groestl);
	sph_groestl512(&ctx_groestl, (uint8_t *)&cnt, sizeof((uint8_t *)&cnt));
	sph_groestl512(&ctx_groestl, input, length);
	sph_groestl512(&ctx_groestl, salt, salt_length);
	sph_groestl512_close(&ctx_groestl, blocks[0]);
	cnt++;

	for (int m = 1; m < s_cost; m++) {
		sph_groestl512(&ctx_groestl, (uint8_t *)&cnt, sizeof((uint8_t *)&cnt));
		sph_groestl512(&ctx_groestl, blocks[m - 1], 64);
		sph_groestl512_close(&ctx_groestl, blocks[m]);
		cnt++;
	}

	// Step 2: Mix buffer contents
	for (uint64_t t = 0; t < t_cost; t++) {
		for (uint64_t m = 0; m < s_cost; m++) {
			// Step 2a: Hash last and current blocks
			sph_groestl512(&ctx_groestl, (uint8_t *)&cnt, sizeof((uint8_t *)&cnt));
			sph_groestl512(&ctx_groestl, blocks[(m - 1) % s_cost], 64);
			sph_groestl512(&ctx_groestl, blocks[m], 64);
			sph_groestl512_close(&ctx_groestl, blocks[m]);
			cnt++;

			for (uint64_t i = 0; i < delta; i++) {
				uint8_t index[64];
				sph_groestl512(&ctx_groestl, (uint8_t *)&t, sizeof((uint8_t *)&t));
				sph_groestl512(&ctx_groestl, (uint8_t *)&cnt, sizeof((uint8_t *)&cnt));
				sph_groestl512(&ctx_groestl, (uint8_t *)&m, sizeof((uint8_t *)&m));
				sph_groestl512(&ctx_groestl, salt, slength);
				sph_groestl512(&ctx_groestl, (uint8_t *)&i, sizeof((uint8_t *)&i));
				sph_groestl512_close(&ctx_groestl, index);
				cnt++;

				uint64_t other = u8tou64(index) % s_cost;
				cnt++;

				sph_groestl512(&ctx_groestl, (uint8_t *)&cnt, sizeof((uint8_t *)&cnt));
				sph_groestl512(&ctx_groestl, block_index(blocks, m), 64);
				sph_groestl512(&ctx_groestl, block_index(blocks, other), 64);
				sph_groestl512_close(&ctx_groestl, block_index(blocks, m));
				cnt++;
			}
		}
	}
	
	memcpy(output, blocks[s_cost - 1], 32);
}
